harness/determined/experimental/client.py

```diff
@@ -0,0 +1,154 @@
+import functools
+import pathlib
+from typing import Any, Callable, Dict, List, Optional, Union
+
+from determined.common.experimental.checkpoint import Checkpoint
+from determined.common.experimental.determined import Determined
+from determined.common.experimental.experiment import ExperimentReference
+from determined.common.experimental.model import Model, ModelOrderBy, ModelSortBy
+from determined.common.experimental.trial import TrialReference
+
+_determined = None  # Optional[Determined]
+
+
+def _require_singleton(fn: Callable) -> Callable:
+    @functools.wraps(fn)
+    def _fn(*args: Any, **kwargs: Any) -> Any:
+        global _determined
+        if _determined is None:
+            _determined = Determined()
+        return fn(*args, **kwargs)
+
+    return _fn
+
+
+def login(
+    master: Optional[str] = None,
+    user: Optional[str] = None,
+    cert_path: Optional[str] = None,
+    cert_name: Optional[str] = None,
+    noverify: bool = False,
+) -> None:
+    """
+    login will configure the default Determined() singleton used by all of the other functions in
+    the client module.
+
+    It is often unnecessary to call login().  If you have configured your environment so that
+    the Determined CLI works without any extra arguments or environment variables, you should not
+    have to call login at all.
+
+    If you do need to call login(), it must be called before any calling any other functions from
+    this module, otherwise it will fail.
+
+    If you have reason to connect to multiple masters, you should use explicit Determined() objects
+    instead.  Each explicit Determined() object accepts the same parameters as login(), and offers
+    the same functions as what are offered in this module.
+    """
+    global _determined
+
+    if _determined is not None:
+        raise ValueError(
+            "You can only call login() once, before any other calls to any functions in the client "
+            "module.  If you have reason to connect to multiple masters, you should use explicit "
+            "client.Determined() objects, which each expose the same functions as this module."
+        )
+
+    _determined = Determined(master, user, cert_path, cert_name, noverify)
+
+
+@_require_singleton
+def create_experiment(
+    config: Union[str, pathlib.Path, Dict],
+    model_dir: str,
+) -> ExperimentReference:
+    """
+    Create an experiment with config parameters and model directory. The function
+    returns :class:`~determined.experimental.ExperimentReference` of the experiment.
+
+    Arguments:
+        config(string, pathlib.Path, dictionary): experiment config filename (.yaml)
+            or a dict.
+        model_dir(string): directory containing model definition.
+    """
+    assert _determined is not None
```

Comment:
I'd say these asserts are redundant. when `_determined` is in fact `None`, next line will throw NoneType AttributeError, which is just as readable and developer-friendly.

Reply:
These asserts are really just for mypy.  They're impossible to reach, because the decorator guarantees they are not hit.